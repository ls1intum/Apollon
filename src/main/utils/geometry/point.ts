export interface IPoint {
  x: number;
  y: number;
}

export class Point implements IPoint {
  constructor(
    public x: number = 0,
    public y: number = 0,
  ) {}

  get length(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  normalize(): Point {
    return new Point(this.x / this.length, this.y / this.length);
  }

  add(x: number, y: number): Point;
  add(x: Point): Point;
  add(x: number | Point, y?: number): Point {
    if (x instanceof Point) {
      return new Point(this.x + x.x, this.y + x.y);
    }
    if (typeof y === 'number') {
      return new Point(this.x + x, this.y + y);
    }
    return this.clone();
  }

  subtract(x: number, y: number): Point;
  subtract(x: Point): Point;
  subtract(x: number | Point, y?: number): Point {
    if (x instanceof Point) {
      return new Point(this.x - x.x, this.y - x.y);
    }
    if (typeof y === 'number') {
      return new Point(this.x - x, this.y - y);
    }
    return this.clone();
  }

  round(radix: number = 10): Point {
    return new Point(Math.round(this.x / radix) * radix, Math.round(this.y / radix) * radix);
  }

  scale(factor: number) {
    return new Point(this.x * factor, this.y * factor);
  }

  clone() {
    return new Point(this.x, this.y);
  }
}
