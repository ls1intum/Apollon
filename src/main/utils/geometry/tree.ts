import { UMLContainer } from '../../services/uml-container/uml-container';
import { IUMLElement, UMLElement } from '../../services/uml-element/uml-element';
import { UMLElementState } from '../../services/uml-element/uml-element-types';

export function filterRoots(ids: string[], elements: UMLElementState): string[] {
  const getSelection = (root: IUMLElement): string[] => {
    if (ids.includes(root.id)) return [root.id];

    if (UMLContainer.isUMLContainer(root)) {
      return root.ownedElements.reduce<string[]>((selection, id) => [...selection, ...getSelection(elements[id])], []);
    }
    return [];
  };

  return Object.values(elements)
    .filter((element) => !element.owner)
    .reduce<string[]>((selection, element) => [...selection, ...getSelection(element)], []);
}

/**
 * returns the ids of all elements in the hierarchy above the element with the specified id
 * @param id element id which parents should be found
 * @param elements elements in state
 */
export function getAllParents(id: string, elements: UMLElementState): string[] {
  const getParents = (element: IUMLElement): string[] => {
    // reached top
    if (element.owner === null) return [];
    return [element.owner, ...getParents(elements[element.owner])];
  };
  return getParents(elements[id]);
}

export function getChildren(ids: string[], elements: UMLElementState): string[] {
  const result: string[] = [];

  for (const id of ids) {
    const owner = elements[id];
    if (!owner) continue;

    if (UMLContainer.isUMLContainer(owner)) {
      result.push(...getChildren(owner.ownedElements, elements));
    }
    result.push(owner.id);
  }

  return result;
}

export function clone(element: UMLElement, elements: UMLElement[]): UMLElement[] {
  if (!UMLContainer.isUMLContainer(element)) {
    return [element.clone()];
  }

  const result: UMLElement[] = [];
  const cloned = element.clone<UMLContainer>();
  const { ownedElements } = element;
  for (const id of ownedElements) {
    const child = elements.find((prev) => prev.id === id);
    if (!child) {
      continue;
    }

    const [clonedChild, ...clonedChildren] = clone(child, elements);
    clonedChild.owner = cloned.id;

    const index = cloned.ownedElements.findIndex((x) => x === id);
    cloned.ownedElements[index] = clonedChild.id;
    result.push(clonedChild, ...clonedChildren);
  }

  return [cloned, ...result];
}
