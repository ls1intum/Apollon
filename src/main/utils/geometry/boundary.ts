import { Point } from './point';

export interface IBoundary {
  x: number;
  y: number;
  width: number;
  height: number;
}

export function computeBoundingBox(points: Point[]): IBoundary {
  if (points.length === 0) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }

  const firstPoint = points[0];

  let minX = firstPoint.x;
  let minY = firstPoint.y;

  let maxX = firstPoint.x;
  let maxY = firstPoint.y;

  for (const p of points) {
    if (p.x < minX) minX = p.x;
    if (p.y < minY) minY = p.y;

    if (p.x > maxX) maxX = p.x;
    if (p.y > maxY) maxY = p.y;
  }

  return {
    x: minX,
    y: minY,
    width: maxX - minX,
    height: maxY - minY,
  };
}

export function computeBoundingBoxForElements(elements: { bounds: IBoundary }[]): IBoundary {
  if (!elements.length) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }
  const boundaries: IBoundary[] = elements.map<IBoundary>((element) => ({ ...element.bounds }));
  const x = Math.min(...boundaries.map((bounds) => bounds.x));
  const y = Math.min(...boundaries.map((bounds) => bounds.y));
  const width = Math.max(...boundaries.map((bounds) => bounds.x + bounds.width)) - x;
  const height = Math.max(...boundaries.map((bounds) => bounds.y + bounds.height)) - y;
  return { x, y, width, height };
}

export function computeDimension(scale: number, value: number, isCircle?: boolean): number {
  if (isCircle && scale === 1) {
    return value * scale;
  } else {
    return Math.round((value * scale) / 10) * 10;
  }
}
