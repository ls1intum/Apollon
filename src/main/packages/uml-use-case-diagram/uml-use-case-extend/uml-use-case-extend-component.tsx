import React, { FunctionComponent } from 'react';
import { Text } from '../../../components/controls/text/text';
import { Point } from '../../../utils/geometry/point';
import { UMLUseCaseExtend } from './uml-use-case-extend';
import { ThemedPath } from '../../../components/theme/themedComponents';

const Arrow: FunctionComponent<{ id: string; color?: string; d: string }> = ({ id, color, d }) => (
  <g>
    <marker
      id={`marker-${id}`}
      viewBox="0 0 30 30"
      markerWidth="22"
      markerHeight="30"
      refX="30"
      refY="15"
      orient="auto"
      markerUnits="strokeWidth"
    >
      <ThemedPath d="M0,29 L30,15 L0,1" fillColor="none" strokeColor={color} />
    </marker>
    <ThemedPath d={d} strokeDasharray={7} markerEnd={`url(#marker-${id})`} strokeColor={color} />
  </g>
);

export const UMLUseCaseExtendComponent: FunctionComponent<Props> = ({ element }) => {
  const [start, end] = element.path.map((p) => new Point(p.x, p.y));
  const line = end.subtract(start);

  if (line.length <= 100) {
    return <Arrow id={element.id} color={element.strokeColor} d={`M ${start.x} ${start.y} L ${end.x} ${end.y}`} />;
  }

  const norm = line.normalize();
  const center = start.add(norm.scale(0.5 * line.length));
  const startSection = start.add(norm.scale(0.5 * line.length - 40));
  const endSection = end.subtract(norm.scale(0.5 * line.length - 40));
  return (
    <g>
      <Arrow
        color={element.strokeColor}
        id={element.id}
        d={`
          M ${start.x} ${start.y} L ${startSection.x} ${startSection.y}
          M ${endSection.x} ${endSection.y} L ${end.x} ${end.y}
        `}
      />
      <path
        id={`textpath-${element.id}`}
        d={`
          M ${startSection.x} ${startSection.y}
          L ${endSection.x} ${endSection.y}
        `}
      />
      <Text
        noX
        noY
        fill={element.textColor}
        transform={
          norm.x < 0
            ? `
              translate(${center.x}, ${center.y})
              rotate(180)
              translate(${-center.x}, ${-center.y})
            `
            : undefined
        }
      >
        <textPath xlinkHref={`#textpath-${element.id}`} startOffset="50%">
          «extend»
        </textPath>
      </Text>
    </g>
  );
};

type Props = {
  element: UMLUseCaseExtend;
};
