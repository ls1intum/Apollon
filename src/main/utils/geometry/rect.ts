import { Direction } from '../../services/uml-element/uml-element-port';
import { IBoundary } from './boundary';
import { Point } from './point';

export function position(bounds: IBoundary, direction: Direction): { point: Point; offset: Point } {
  const { x, y, width, height } = bounds;
  const offset = 40;
  switch (direction) {
    case Direction.Up:
      return {
        point: new Point(x + width, y),
        offset: new Point(x + width, y - offset),
      };
    case Direction.Right:
      return {
        point: new Point(x + width, y + height),
        offset: new Point(x + width + offset, y + height),
      };
    case Direction.Down:
      return {
        point: new Point(x + width, y + height),
        offset: new Point(x + width, y + height + offset),
      };
    case Direction.Left:
      return {
        point: new Point(x, y + height),
        offset: new Point(x - offset, y + height),
      };
  }
}
