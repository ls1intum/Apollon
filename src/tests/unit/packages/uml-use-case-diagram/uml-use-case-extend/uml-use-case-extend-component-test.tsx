import * as React from 'react';
import { wrappedRender } from '../../../test-utils/render';
import { Point } from '../../../../../main/utils/geometry/point';
import { UMLUseCaseExtend } from '../../../../../main/packages/uml-use-case-diagram/uml-use-case-extend/uml-use-case-extend';
import { UMLUseCaseExtendComponent } from '../../../../../main/packages/uml-use-case-diagram/uml-use-case-extend/uml-use-case-extend-component';

it('render the uml-use-case-extend-component', () => {
  const umlUseCaseExtend: UMLUseCaseExtend = new UMLUseCaseExtend({
    id: '225bb7a6-5af4-4473-ba96-7a30b02545c4',
    path: [new Point(0, 0), new Point(100, 100)],
  });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <UMLUseCaseExtendComponent element={umlUseCaseExtend} />
    </svg>,
  );
  expect(getByText('«extend»')).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
