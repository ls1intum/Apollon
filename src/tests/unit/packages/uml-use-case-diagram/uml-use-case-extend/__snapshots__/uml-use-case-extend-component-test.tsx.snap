// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the uml-use-case-extend-component 1`] = `
<body>
  <div>
    <svg>
      <g>
        <g>
          <marker
            id="marker-225bb7a6-5af4-4473-ba96-7a30b02545c4"
            markerHeight="30"
            markerUnits="strokeWidth"
            markerWidth="22"
            orient="auto"
            refX="30"
            refY="15"
            viewBox="0 0 30 30"
          >
            <path
              class="sc-gsDKAQ datHse"
              d="M0,29 L30,15 L0,1"
              fill="none"
              stroke="black"
            />
          </marker>
          <path
            class="sc-gsDKAQ ebiVBv"
            d="
          M 0 0 L 21.715728752538098 21.715728752538098
          M 78.2842712474619 78.2842712474619 L 100 100
        "
            fill="white"
            marker-end="url(#marker-225bb7a6-5af4-4473-ba96-7a30b02545c4)"
            stroke="black"
            stroke-dasharray="7"
          />
        </g>
        <path
          d="
          M 21.715728752538098 21.715728752538098
          L 78.2842712474619 78.2842712474619
        "
          id="textpath-225bb7a6-5af4-4473-ba96-7a30b02545c4"
        />
        <text
          dominant-baseline="middle"
          font-weight="bold"
          pointer-events="none"
          text-anchor="middle"
        >
          <textpath
            startOffset="50%"
            xlink:href="#textpath-225bb7a6-5af4-4473-ba96-7a30b02545c4"
          >
            «extend»
          </textpath>
        </text>
      </g>
    </svg>
  </div>
</body>
`;
